### DISCUSSION TITLE
Passing multiple costing modes to the /centroid endpoint

### DISCUSSION THREAD

**Discussion by LennartKoepper** [valhalla/valhalla#4621] [Q&A] [answered]
Is there a way to pass multiple costing modes associated to the different starting locations to the /centroid endpoint? 
Use-Case Example: Determining the optimal meeting point for multiple road-users of different types, e.g two cyclists and a car, given their current location.

**Comment by kevinkreiser** [comment, marked as answer]
interesting question. sadly no, that feature currently doesnt exist and implementing it is unfortunately non trivial. the reason for that is two fold:

1. the way the expansion is done search paths can and do collapse into a single path to be expanded in the queue. the problem with different costing models is that the algorithm could no longer work this way because different costings will create different sort costs and thus need their own queues. the queues only really work if the things in them are measured in the same "units". anyway we could have a different queue for each costing model but this would be both mega inefficient and horrible to maintain
2. different costings, as i mentioned above, measure the weight of edges/nodes differently. most times they are a function of distance with some other fixed penalties but also some penalty multipliers. what this means is that bike costing might think a given edge is 20 units of cost whereas a car thinks its 3 units of cost. the bike and the car have a different world view, a different notion about what a good vs bad path is. numerically this makes a centroid computation not well defined. as it exists today we seek to find a point that minimizes the cost of all paths to get there, similar to the concept of a geometric centroid (also schwerpunkt) but in this case the metric isnt distance, its graph traversal cost. we can do this minimization because any given edge/node/path is sortable/comparable against any other and this is because all of them have the same units, either bike units, or car units or truck units, whatever.

so yeah number 2 there is really the key problem: it currently doesnt make sense to compare cost values from different costings to each other. however this **could** technically be overcome and i think we have two options for that:

1. make it so that different costings have the same magnitude of cost values after controlling for the length of the edge and the speed difference in the modes of travel. basically no cost can have a value that is `f(d) = max_scaling * d + max_penalty`. if all costings shared the same cap on the scaling penalty and the constant penalties then the magnitude of the cost values would be gauranteed to be the same. the main problem here is that speed of travel affects the scaling and so cars will consistently have cheaper costs than ped and bike though this doesnt seem unfair actually if we control the other factors appropriately it would just mean the centroid would favor more driving than walking and biking.
2. which leads me to the other option, we could simply use estimated travel time as the metric to minimize and since it is consistent as a unit across all costings it would work without modification to the costing algorithms (ie its much easier to implement) the downside is we'd miss out on all the heuristic parts that add time penalty to the cost but not to the ETA, so the routes would be a bit strange.

ok yeah i hope that gives a flavor of the difficulties in taking on something like that. it seems there are some ugly things we could try but the main issue is, its hard for the algorithm to tell what it should expand next if the list of options are not comparable to each other.
